File: src/main/kotlin/com/vlnabatov/alabaster/extensions/annotation/KotlinAnnotator.kt
package com.vlnabatov.alabaster.extensions.annotation

import com.intellij.lang.annotation.AnnotationHolder
import com.intellij.lang.annotation.Annotator
import com.intellij.lang.annotation.HighlightSeverity.INFORMATION
import com.intellij.openapi.editor.DefaultLanguageHighlighterColors.BRACES
import com.intellij.openapi.editor.DefaultLanguageHighlighterColors.NUMBER
import com.intellij.psi.PsiElement
import com.intellij.psi.impl.source.tree.LeafPsiElement
import com.intellij.psi.util.elementType
import com.vlnabatov.alabaster.annotateSeparationMarks
import org.jetbrains.kotlin.psi.stubs.elements.KtStubElementTypes.*


class KotlinAnnotator : Annotator {
    override fun annotate(element: PsiElement, holder: AnnotationHolder) {
        try {
            when (element.elementType) {
                NULL, BOOLEAN_CONSTANT, FLOAT_CONSTANT, CHARACTER_CONSTANT, INTEGER_CONSTANT ->
                    holder.newSilentAnnotation(INFORMATION).textAttributes(NUMBER).create()

                STRING_TEMPLATE ->
                    annotateSeparationMarks(
                        element,
                        holder,
                        BRACES,
                        (element.firstChild as LeafPsiElement).cachedLength
                    )
            }
        } catch (e: Exception) {
            /* Should not happen */
        }
    }
}

File: src/main/clojure/com/vlnabatov/alabaster/kotlin_annotator.clj
(ns com.vlnabatov.alabaster.kotlin-annotator
  (:import (com.intellij.lang.annotation AnnotationHolder)
           (com.intellij.lang.annotation HighlightSeverity)
           (com.intellij.openapi.editor DefaultLanguageHighlighterColors)
           (com.intellij.psi PsiElement)
           (com.intellij.psi.tree IElementType)
           (org.jetbrains.kotlin.psi.stubs.elements KtStubElementTypes)
           (com.vlnabatov.alabaster UtilKt))
  (:gen-class :implements [com.intellij.lang.annotation.Annotator]
              :name       com.vlnabatov.alabaster.extensions.annotation.KotlinAnnotator
              :prefix     "-"))


(defn -isDumbAware [_this] false)

(defn -annotate
  [_ ^PsiElement element ^AnnotationHolder holder]
  (let [^IElementType elementType (.. element getNode getElementType)]
    (cond (= KtStubElementTypes/STRING_TEMPLATE elementType) (do (UtilKt/annotateSeparationMarks element holder))
          (contains? #{KtStubElementTypes/NULL KtStubElementTypes/BOOLEAN_CONSTANT KtStubElementTypes/FLOAT_CONSTANT
                       KtStubElementTypes/CHARACTER_CONSTANT KtStubElementTypes/INTEGER_CONSTANT}
                     elementType)
            (do (-> holder
                    (.newSilentAnnotation HighlightSeverity/INFORMATION)
                    (.textAttributes DefaultLanguageHighlighterColors/NUMBER)
                    (.create))))))